src/type-layout.md

```diff
@@ -236,8 +236,12 @@ the sake of clarity. To perform memory layout computations in actual code, use
 
 </div>
 
-> Note: This algorithm can produce zero-sized structs. This differs from
-> C where structs without data still have a size of one byte.
+> Note: This algorithm can produce zero-sized structs. In C, an empty struct
+> declaration like `struct Foo { }` is illegal. However, both gcc and clang
+> support options to enable such structs, and assign them size zero. C++, in
+> contrast, gives empty structs a size of 1, unless they are inherited from or
+> they are fields that have the `[[no_unique_address]]` attribute, in which
+> case they do not increase the overall size of the struct.
```

Comment:
The last bit (empty base optimization / ``[[no_unique_address]]``) is not mentioned in UCG and seems dubiously relevant: `sizeof(Foo)` is not affected by those two tricks, they just modify the layout computation of *other* types. Those other types explicitly do things that don't have a Rust equivalent (inheritance / said attribute), so I don't think this will be as relevant as the rest of this note for people translating struct definitions from C++ to Rust.

Reply:
@hanna-kruppe The ``[[no_unique_address]]` is mentioned in UCG I guess? Actually the EBO is not limited to inheritance: 
<https://en.cppreference.com/w/cpp/language/ebo>
 ```cpp
#include <cassert>

struct Empty {}; // empty class
 
struct X {
    int i;
    [[no_unique_address]] Empty e;
};
 
int main()
{
    // the size of any object of empty class type is at least 1
    assert(sizeof(Empty) >= 1);
 
    // empty member optimized out
    assert(sizeof(X) == sizeof(int));
}
```

Reply:
Oh, you're right, both empty base optimization and `[[no_unique_address]]` are mentioned in UCG, sorry for the confusion. I also take issue with the wording in UCG (and I can't see how your link or example supports that "EBO is not limited to inheritance" -- the attribute is separate from empty *base* optimization), but that need not block this PR.